### PR TITLE
[TransformStrategies] Add support for aligned and partially aligned matmul

### PR DESCRIPTION
**Pull request by qedawkins**
Current handling of tile aligned matmuls differs from unaligned in the way the tensor.pad ops are lowered. When the pad is aligned and would otherwise fold, we convert to a linalg.copy and don't need to do masked vectorization. In the future the plan is to unify aligned and unaligned cases.

Currently due to some observed performance regressions, any alignment is guarded behind `--iree-codegen-llvmgpu-enable-transform-dialect-aligned-matmul`.

### CODE REVIEW

**Comment by qedawkins** [comment]
Rebase to rerun broken Github action from earlier.

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp:863]: There is a concurrent effort to activate the unaligned case by default.

Could we keep the separation here and introduce a new flag (off by default) to activate the aligned case ?

I'd like to keep the 2 concerns separate because I suspect there is some missing stuff in the aligned case vs the current IREE pipeline.

**Comment by nicolasvasilache** [comment]
Generally looks great but let's please put the aligned case behind a new flag to avoid interfering with the activation of the unaligned case by default.

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp:863]: Sounds good. Just trying a few benchmarks reveals that aligned here is a little behind default IREE (even picking the same tile sizes) so will add an additional flag.

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp:863]: Partially aligned seems to still be better than the current default SIMT path so I left that out of the flag. (e.g. 650µs vs 2ms on A100 in the example in the tests with whatever tile sizes are picked by default).

Although my benchmarking was fairly ad-hoc so would be curious to know how others are doing it. I didn't see how to do it from the dispatch profiler at least.

**Comment by MaheshRavishankar** [comment]
When you say enable, do you mean enable by default? I dont see that in the code (making sure I dont miss anything)

**Comment by qedawkins** [comment]
> When you say enable, do you mean enable by default? I dont see that in the code (making sure I dont miss anything)

Oh I didn't mean that. Updated the title accordingly.

**Comment by qedawkins** [comment]
> When you say enable, do you mean enable by default? I dont see that in the code (making sure I dont miss anything)

For a bit more context, there is a parallel effort to enable these strategies by default for unaligned cases (#13492), however the fully aligned case will still remain off by default with the way things stand now. Partially aligned will also be turned on at the same time (my understanding, others can correct me if otherwise).

**Comment by qedawkins** [comment]
@nicolasvasilache Gentle bump here; is there anything else blocking on this and/or that we want to wait on?

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp]: Are these guards actually necessary?
This should do nothing in case there are no masks right ?
If possible, I'd rather keep simplicity here.

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp]: Can you add a comment as to why this is necessary?
I'll try and remove it in a followup.

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp]: can we also keep this unconditional?
Nothing should happen when unneeded

**Review by nicolasvasilache — APPROVED** [review]
Thanks @qedawkins and sorry for the delay, I had some local challenges that impacted my schedule.
Generally this looks good to me, this will require a few refactoring followups but it is great you were able to unify the 2 cases already.

Some minor comments around the fact that I'd rather reduce conditional complexity and let the transforms "do nothing" where applicable.

**Comment by nicolasvasilache** [comment]
@qedawkins can you please rebase on #13492 and check there are no regressions before landing ?

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp]: (Leaving a longer comment here first)

First for rematching the fill, I see that I can just use the fill handle above from block distribution so that part is unnecessary and I will drop it with the rebase.

The important part here is running canonicalization. For aligned right now we are relying on canonicalization to simplify the `tensor.pad` such that once we rewrite in DPS we get a `linalg.copy`. For example, without canonicalization we will have something like:
```
%padded = tensor.pad %extracted_slice_8 nofold low[%c0_12, %c0_12] high[%14, %15] {
^bb0(%arg5: index, %arg6: index):
   tensor.yield %cst_11 : f16
} : tensor<128x16xf16> to tensor<128x16xf16>
```
and after DPS
```
%17 = linalg.fill ins(%cst_16 : f16) outs(%16 : tensor<128x16xf16>) -> tensor<128x16xf16>
%inserted_slice = tensor.insert_slice %extracted_slice_8 into %17[%c0_12, %c0_12] [128, 16] [1, 1] : tensor<128x16xf16> into tensor<128x16xf16>
```
Whereas if we do canonicalization between the pad and DPS then we get
```
%padded = tensor.pad %extracted_slice_2 nofold low[%c0, %c0] high[%c0, %c0] {
^bb0(%arg5: index, %arg6: index):
  tensor.yield %cst : f16
} : tensor<128x16xf16> to tensor<128x16xf16>
```
into
```
%11 = bufferization.alloc_tensor() : tensor<128x16xf16>
%12 = linalg.copy ins(%extracted_slice_2 : tensor<128x16xf16>) outs(%11 : tensor<128x16xf16>) -> tensor<128x16xf16>
```
(also annoyingly if we do canonicalization after rewriting to dps instead of before all the copies/pads get folded away)
```
%extracted_slice_2 = tensor.extract_slice %extracted_slice[0, %arg3] [128, 16] [1, 1] : tensor<128x1024xf16> to tensor<128x16xf16>
%extracted_slice_3 = tensor.extract_slice %extracted_slice_0[%arg3, 0] [16, 128] [1, 1] : tensor<1024x128xf16> to tensor<16x128xf16>
%11 = linalg.matmul ins(%extracted_slice_2, %extracted_slice_3 : tensor<128x16xf16>, tensor<16x128xf16>) outs(%arg4 : tensor<128x128xf16>) -> tensor<128x128xf16>
```

Right now, `HoistOutputPaddingOp` runs canonicalization as a last step, but we shouldn't do hoisting if the result is unaligned. Also IIRC if we canonicalize before hoisting then there is a chance that the extract_slice on the loop variable gets folded out causing hoisting to fail (definitely some brittleness to address there) so canonicalizing after the pad isn't an option. To summarize

1. If any pads are aligned then we must canonicalize before converting to DPS to get the copy
2. If the output is unaligned, we will run canonicalization anyway so aligned input pads will get the canonicalization they need anyway.
3. Canonicalization cannot run between padding and hoisting and we do not want to hoist an aligned output.

I can just make canonicalization unconditionally run after conditionally doing the output hoisting to remove the `else`, although I figured it would be ugly to run canonicalization back-to-back in the unaligned case. WDYT?

**Comment by qedawkins** [comment]
Looking at the CI failures I'm seeing a few problems. First, the accuracy error in the e2e tests I added might not be isolated to the aligned case; I filed #13811 for the correctness issues with mma.sync that I noticed. Additionally trying `512x500x128` with this PR + WMMA seems to only give incorrect results when async copies are enabled, but I haven't yet seen the same problem for fully unaligned so still investigating.

Second, it seems that one of the real-world shapes `77x640x768` revealed an issue that I thought was only present for fully aligned, but essentially canonicalizing after tiling to `scf.for` can cause output hoisting to fail for unaligned as well due to the extract_slice being folded away. I'm going to switch to never canonicalizing for now to help avoid this for now.

**Comment by MaheshRavishankar** [comment]
cc @qedawkins and @manishucsd . I think this is related to some of the issues we've seen here so far...

**Comment by qedawkins** [comment]
Is there an easy way to get the IR for the models in the benchmark report?

EDIT: I found the link to the IR in the test

**Comment by MaheshRavishankar** [comment]
> Is there an easy way to get the IR for the models in the benchmark report?

https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/benchmark_suites.md has some instructions. Typically in the output log if you see the `https://...` path to the benchmark file you can do `wget` to get the input file.

**Comment by MaheshRavishankar** [comment]
> Is there an easy way to get the IR for the models in the benchmark report?
> 
> EDIT: I found the link to the IR in the test

Ah ok, I was just going to say you can use `wget https://storage.googleapis.com/iree-model-artifacts/pytorch/torch_models_20230307.103_1678163233/SD_CLIP_TEXT_MODEL_SEQLEN64/linalg.mlir` ... but looks like you found it.

**Comment by qedawkins** [comment]
The above regressions are a known issue so the plan for the time being is to keep cases with any aligned dimensions and flip the switch at least for partially aligned once resolved. I'm planning to land this tomorrow once CI is passing in this case, assuming no other blockers.

**Pull request merged**
